PTP-only documentation improvements (#1188)
suse-edge/suse-edge.github.io: asciidoc/product/atip-features.adoc

* *Synchronous Ethernet (SyncE)*: physical-layer frequency synchronization. PTP
carries phase and time inside packets, whereas SyncE recovers a stable frequency
directly from the Ethernet signal. SyncE is an *optional enhancement*: PTP on its
own already delivers phase and time, and SyncE is added only when a node needs
stronger frequency stability and longer holdover should the PTP time source be
temporarily lost.
* *Digital Phase-Locked Loop (DPLL)*: a hardware timing block that locks onto a
reference signal and filters it to output a stable, clean clock. It is what
disciplines the PHC from the available inputs and, on platforms with more than one
NIC, keeps the different NIC clocks aligned.

On each node these functions are carried out by the `linuxptp` daemons: `ptp4l`
runs the PTP protocol and disciplines the PHC, and `phc2sys` aligns the system
clock to the PHC. On a Grand Master, `ts2phc` additionally steers the PHC from the
GNSS reference. Their roles and installation are covered in <<ptp-prerequisites>>.

The hardware clock itself is handled by the Linux kernel DPLL subsystem together
with the network driver. How much of the timing can be controlled depends on

> whether the NIC integrates a manageable DPLL:

Reviewer: ```suggestion
with the network driver. The degree of timing control depends on whether the NIC integrates a manageable DPLL.:
```

Author: Done. Kept a single colon (`...manageable DPLL:`) since a bulleted list follows.

[#ptp-configuration]
== Precision Time Protocol

[#ptp-introduction]
=== Introduction

Precision Time Protocol (PTP), defined by the IEEE 1588 standard, distributes
highly accurate time across a packet network. When the network interface cards
(NICs) timestamp PTP packets in hardware, it reaches sub-microsecond accuracy,
well beyond what NTP can offer. In telco environments PTP is a core building
block of 5G radio access networks (RAN), where the radio units require tight
phase and time alignment to transmit correctly. SUSE Telco Cloud implements PTP
through the `linuxptp` package and targets the telco profiles ITU-T G.8275.1 and
ITU-T G.8275.2 described below.

A PTP domain is built from a small set of clock roles:

* *Grandmaster (GM)*: the primary source of time for the domain. It is normally
locked to a Primary Reference Time Clock (PRTC), usually driven by a GNSS
receiver, and transmits time towards the downstream nodes. The accuracy of the GM
sets the quality ceiling for the whole domain.
* *Ordinary Clock (OC)*: an end node with a single PTP port. In a telco deployment
it acts as a Time Receiver, disciplining its local clock from an upstream Grand

> Master or Boundary Clock. A compute node running a radio workload is a typical OC.

Reviewer 2: AFAIK, the typical scenario is a compute node is running as a BC, since timing sync needs to be provided to the RU(s).

        all
----

Check for the `hardware-transmit`, `hardware-receive` and `hardware-raw-clock`
capabilities and a `PTP Hardware Clock` index of 0 or higher (a value of `-1` means
the interface has no PHC). Replace `p1p1` with the interface you intend to use for
PTP.

PTP is provided by the `linuxptp` package. It ships three user-space programs —
`ptp4l`, `phc2sys` and `ts2phc` — that run as long-running daemons managed by

> `systemd` (a dedicated service each: `ptp4l.service`, `phc2sys.service` and

Reviewer: ```suggestion
`systemd` (a dedicated service each: `ptp4l.service`, `phc2sys.service`, and
```

Author: Done — added the Oxford comma.

`ts2phc.service`). Together they run the protocol and keep the different clocks

> aligned:

Reviewer: ```suggestion
`ts2phc.service`). Collectively, they execute the protocol and keep the clocks synchronized: 
```

Author: Done.

  baseImage: {micro-base-rt-image-raw}
  outputImageName: eibimage-slmicrort-telco.raw
operatingSystem:
  systemd:
    enable:
      - ptp4l
      - phc2sys
  packages:
    packageList:
      - linuxptp
    sccRegistrationCode: $SCC_REGISTRATION_CODE
----

[NOTE]
====

> The `linuxptp` package does not enable `ptp4l`, `phc2sys` or `ts2phc` by default.

Reviewer: ```suggestion
The `linuxptp` package does not enable `ptp4l`, `phc2sys`, or `ts2phc` by default.
```

Author: Done — added the Oxford comma.

Enable the daemons required by each node through the `systemd` section of the

> manifest, once their configuration files are in place (see the scenarios below).

Reviewer: ```suggestion
After deploying the configuration files, use the `systemd` section of the manifest to enable the required node daemons(see the scenarios below).
```

Author: Done (also added the missing space: `daemons (see ...`).

below the BC in the BMCA:
+
[#ptp-conf-bc-single-nic-oc]
.Ordinary Clock configuration
====
[source,]
----
# Telecom G.8275.1 — Ordinary Clock (Time Receiver)
[global]
domainNumber                    24
priority2                       255
dataset_comparison              G.8275.x
G.8275.portDS.localPriority     250
G.8275.defaultDS.localPriority  250
maxStepsRemoved                 255
logAnnounceInterval             -3
logSyncInterval                 -4
logMinDelayReqInterval          -4
announceReceiptTimeout          3
clientOnly                      1
ptp_dst_mac                     01:80:C2:00:00:0E
network_transport               L2
summary_interval                3
message_tag                     "ptp4l"
uds_address                     /var/run/ptp4l
uds_ro_address                  /var/run/ptp4lro

[ens6f0np0]
----
====

. Configure `/etc/sysconfig/ptp4l` and `/etc/sysconfig/phc2sys` as for the BC (the
`phc2sys` source is the PHC of `ens6f0np0`), then enable and start the services:
+
[source,]
----
# /etc/sysconfig/ptp4l
OPTIONS="-f /etc/ptp4l-G.8275.1.conf"

# /etc/sysconfig/phc2sys
OPTIONS="-s /dev/ptp1 -w -m --uds_address /var/run/ptp4l --domainNumber 24"
----

> . Verify that `ens6f0np0` locks to the time regenerated by the BC. The port moves to

Reviewer: ```suggestion
. Verify that `ens6f0np0` is locked to the regenerated time from the BC. The port moves to
```

Author: Done.

equipped with a dedicated DPLL that routes the timing signals across the system and the
different NICs internally, and no longer require external cabling. The overall deployment
is thus simplified, but specific steps and software components are required, such as
boot-time DPLL and internal signal configuration.

This scenario extends <<ptp-scenario-bc-single-nic,Scenario 3>> to a Boundary Clock (BC)
built on several NICs, entirely over ITU-T G.8275.1 (PTP-only, Layer 2). One port of the
integrated controller receives PTP from the Grand Master and the add-in cards regenerate
PTP downstream. Unlike the single-NIC case, each NIC has its own PHC; the on-board DPLL
and `ts2phc` keep them aligned, and one `ptp4l` instance runs per PHC.

Before going through the setup steps, download and install the latest kernel driver for
the _Intel 800 Series Network Devices_ from the
https://www.intel.com/content/www/us/en/download/19630/intel-network-adapter-driver-for-800-series-devices-under-linux.html[Intel website].
Select the zip file containing the "ice_RPM_Files", unpack the archive and install the
`ice-kmp-default-*sles16sp0.x86_64.rpm` RPM it contains, for example

> `ice-kmp-default-2.4.5_k6.12.0_160000.5-1.sles16sp0.x86_64.rpm`:

Reviewer: ```suggestion
.Prerequisites
* Download and install the latest kernel driver for the _Intel 800 Series Network Devices_ from the
https://www.intel.com/content/www/us/en/download/19630/intel-network-adapter-driver-for-800-series-devices-under-linux.html[Intel website].
* Select the zip file containing the "ice_RPM_Files", unpack the archive and install the `ice-kmp-default-*sles16sp0.x86_64.rpm`. This RPM contains, for example, `ice-kmp-default-2.4.5_k6.12.0_160000.5-1.sles16sp0.x86_64.rpm`. 
```

Author: Done — converted to a `.Prerequisites` list. Attached the `transactional-update` command and the reboot line to the second item so they remain part of that prerequisite.

it is the port used to configure the DPLL signals, and it is not necessarily the port that
carries PTP (`em2`).
====

[NOTE]
====
The setup described here is completely static. If the GM becomes unavailable, downstream
synchronization relies exclusively on the host holdover capabilities; no automatic
switching to a different time reference is performed.
====

*Diagram*

.Boundary Clock (PTP-only), multi-NIC GNR-D node with on-board DPLL, with propagation to an Ordinary Clock
image::product-atip-ptp-scenario-bc-multi-nic-dpll.png[Scenario 4 - Boundary Clock (PTP-only) multi-NIC with internal timing module (DPLL GNR-D based),scaledwidth=100%]

*Configuration*

Every switch port on the timing path must have PTP enabled. The BC server is then brought
up in the order below. Each runtime step is also mapped to a `systemd` unit so that the
configuration survives reboots and can be automated with Cluster API (see
<<add-telco-ptp-gnrd,the GNR-D CAPI automation>>). The OC server is configured exactly as in
<<ptp-scenario-bc-single-nic,Scenario 3>>.

. Identify the PTP devices and the integrated controller.
+

> List the PTP clocks and the PCI device each is bound to:

Reviewer: ```suggestion

.. List the PTP clocks and the PCI device each is bound to:
```

Author: Done — converted step 1 into numbered sub-steps (a/b/c). This is sub-step a.

+
[,console]
----
# ls -l /sys/class/ptp/
lrwxrwxrwx 1 root root 0 ... ptp1 -> ../../devices/pci0000:12/0000:12:04.0/0000:13:00.0/ptp/ptp1
lrwxrwxrwx 1 root root 0 ... ptp2 -> ../../devices/pci0000:6b/0000:6b:02.0/0000:6c:00.0/ptp/ptp2
lrwxrwxrwx 1 root root 0 ... ptp3 -> ../../devices/pci0000:6b/0000:6b:06.0/0000:6e:00.0/ptp/ptp3
----
+
Look up each device type with `lspci` to tell the integrated controller apart from the

> add-in cards:

Reviewer: ```suggestion
.. Run `lspci` to identify each device type, allowing you to distinguish between the integrated controller and the add-in cards.
```

Author: Done — now sub-step b of the numbered list.

6e:00.0 Ethernet controller: Intel Corporation Ethernet Controller E830-CC for SFP
----
+
This reference server has an _Intel E825-C_ integrated controller (`/dev/ptp1`) and two
_E830-CC_ add-in cards (`/dev/ptp2`, `/dev/ptp3`). Identify the administrative port of the

> integrated controller, which manages its PHC:

Reviewer: ```suggestion
_E830-CC_ add-in cards (`/dev/ptp2`, `/dev/ptp3`). 

.. Identify the administrative port of the integrated controller, which manages its PHC:
```

Author: Done — split into its own sub-step (c).

echo 4 1 > /sys/class/net/em5/device/tspll_cfg

# Feed the DPLL from the integrated PHC through the SDP pins:
# SDP0 -> 1 pulse per second (1PPS, 1 Hz): enable the pin, then set a 1 s period
echo 2 2 > /sys/class/ptp/ptp1/pins/SDP0
echo 2 0 0 1 0 > /sys/class/ptp/ptp1/period

# SDP2 -> 1000 pulses per second (1kPPS, 1 kHz): enable the pin, then set a 1 ms period
echo 2 1 > /sys/class/ptp/ptp1/pins/SDP2
echo 1 0 0 0 1000000 > /sys/class/ptp/ptp1/period

# Propagate the pulses from the DPLL to the add-in cards (SDP1 on each add-in PHC)
echo 1 1 > /sys/class/ptp/ptp2/pins/SDP1
echo 1 1 > /sys/class/ptp/ptp3/pins/SDP1
----
====
+
These settings are lost on reboot, so run the script at boot from a oneshot `systemd` unit,
ordered before `ts2phc` and `ptp4l`:
+
[source,ini]
----
# /etc/systemd/system/gnrd-ptp-dpll-setup.service
[Unit]
Description=GNR-D PTP DPLL and internal signal setup
Before=ts2phc.service ptp4l@nac.service ptp4l@aic1.service ptp4l@aic2.service

[Service]
Type=oneshot
RemainAfterExit=yes
ExecStart=/usr/local/bin/01-gnrd-startup-setup.sh

[Install]
WantedBy=multi-user.target
----

. Align the add-in cards' PHCs with `ts2phc`.
+

> `ts2phc` ("time stamp to PHC") aligns the add-in cards' PHCs to a reference clock. Here the

Reviewer: ```suggestion
`ts2phc` ("time stamp to PHC") aligns the add-in cards' PHCs to a reference clock. Here, the
```

Author: Done — added the comma after `Here`.

reference is the integrated PHC (`/dev/ptp1`, already disciplined by the GM through its
`ptp4l` instance), passed on the command line with `-s /dev/ptp1`. The on-board DPLL
delivers the 1PPS derived from that reference to each add-in card on its `SDP1` pin;
`ts2phc` timestamps those input pulses (EXTTS) and steps each add-in PHC accordingly, so
every PHC in the node ends up sharing the same time. Create its configuration file:
+
[#gnrd-ts2phc-config]

> .`/etc/ts2phc-cf-all.cfg`

Reviewer: Is this a file? Can you please label the example more clearly?

Author: Yes, it is a file. Labeled the block title `/etc/ts2phc-cf-all.cfg — ts2phc configuration file`, matching the pattern used for the ptp4l config blocks.

ordered before `ts2phc` and `ptp4l`:
+
[source,ini]
----
# /etc/systemd/system/gnrd-ptp-dpll-setup.service
[Unit]
Description=GNR-D PTP DPLL and internal signal setup
Before=ts2phc.service ptp4l@nac.service ptp4l@aic1.service ptp4l@aic2.service

[Service]
Type=oneshot
RemainAfterExit=yes
ExecStart=/usr/local/bin/01-gnrd-startup-setup.sh

[Install]
WantedBy=multi-user.target
----

. Align the add-in cards' PHCs with `ts2phc`.
+
`ts2phc` ("time stamp to PHC") aligns the add-in cards' PHCs to a reference clock. Here, the
reference is the integrated PHC (`/dev/ptp1`, already disciplined by the GM through its
`ptp4l` instance), passed on the command line with `-s /dev/ptp1`. The on-board DPLL
delivers the 1PPS derived from that reference to each add-in card on its `SDP1` pin;
`ts2phc` timestamps those input pulses (EXTTS) and steps each add-in PHC accordingly, so
every PHC in the node ends up sharing the same time.
+
[NOTE]
====
Aligning the add-in PHCs takes two complementary settings, at different levels:

* *Hardware*: the `SDP1` input pin of each add-in PHC must be enabled so the on-board DPLL

> can deliver the 1PPS pulse to it. This is done in the DPLL setup script of the previous step

Reviewer: ```suggestion
* *Hardware*: the `SDP1` input pin of each add-in PHC must be enabled for the on-board DPLL to deliver the 1PPS pulse to it. This is done in the DPLL setup script of the previous step
```

(`echo 1 1 > /sys/class/ptp/ptpN/pins/SDP1`).
* *Software*: `ts2phc` must be told to timestamp that input and step the PHC. This is what the
configuration below does (`ts2phc.channel`, `ts2phc.pin_index`, `ts2phc.extts_polarity`).

Without the pin enablement the pulse never reaches the PHC; without the `ts2phc` configuration

> the pulse is never applied to it. Both are required.

Reviewer: ```suggestion
Without enabling the pin, the pulse never reaches the PHC; without the `ts2phc` configuration
the pulse is never applied to it. Both are required.
```

BMCA to assign each role. With a GM reachable through one of them, the BMCA selects that port
(`em2`) as the Time Receiver (`SLAVE`) and disciplines `/dev/ptp1`, while the remaining ports
follow the BMCA outcome. This is what makes the node a Boundary Clock rather than a plain
receiver, so the ports must stay unconstrained: forcing `clientOnly` would stop it from
forwarding time, and forcing `serverOnly` would stop it from locking to the GM. The add-in
cards, in contrast, force `serverOnly 1` because they must always be Time Transmitters — they
take their time from the DPLL and `ts2phc`, not by receiving PTP, so they must never be
elected as receivers. If tighter control is needed, add the `serverOnly` or `clientOnly` flag
explicitly under each interface.
+
The `G.8275.portDS.localPriority` and `G.8275.defaultDS.localPriority` values (`128` on the
integrated instance, `250` on the add-in cards) act as a G.8275.x BMCA tie-breaker: they are
compared only after `clockClass`, `clockAccuracy`, variance and `priority2`, and a lower value
is preferred. Keeping the default `128` on the integrated instance makes it outrank the add-in

> instances (`250`), which keeps the master/receiver ordering within the node deterministic and

Reviewer: Can master/received be changed to anything else? Just checking because master/slave is not used unless it is for legal protocol.

Author: Good catch. Replaced `master/receiver` with the standard IEEE 1588 roles already used in this doc: `Time Transmitter/Time Receiver`. Kept `GM`, Grandmaster, BMCA and the literal `ptp4l` `best master clock` output, which are protocol/standard terms. Also unified `Grand Master` → `Grandmaster` (single word) across the whole doc and diagrams.

====
The values applied with `pmc` are held only in the running `ptp4l` process; they are *not*
written to the `.conf` file and do *not* survive a restart. If a `ptp4l` instance is
restarted, it reverts to the file defaults and the values must be forwarded again. Automate
the forwarding (see below) rather than running it once by hand.
====
+
Once the integrated controller is locked to the GM, query the grandmaster values. The
helper below takes the configuration file of the instance to query as its argument:
+
[#gnrd-get-pmc-values]
.`get_pmc_values.sh`
====
[source,bash]
----
#!/bin/bash
# Retrieve Grand Master information via pmc for the given ptp4l config file
set -euo pipefail

echo "PARENT_DATA_SET (grandmaster clock values):"
pmc -f "$1" -u -b 0 'GET PARENT_DATA_SET' | grep gm

echo "TIME_PROPERTIES_DATA_SET:"
pmc -f "$1" -u -b 0 'GET TIME_PROPERTIES_DATA_SET'
----
====
+
Then forward those values to the add-in instances. The following script reads the values
from the integrated instance and applies them to each add-in instance. Edit the "Input
variables" section to match your configuration file names, and run it once the system is
synchronized to a GM:
+
[#gnrd-pmc-forward]

> .`04-ptp-forward-GM.sh`

Reviewer: Can you please label this example clearly?

Author: Done — labeled the script blocks with a description: `get_pmc_values.sh — query the Grandmaster values...` and `04-ptp-forward-GM.sh — forward the Grandmaster values...`.

    fi
done
----
====
+
Because the values are not persistent, run this script from a `systemd` timer so that they
are refreshed periodically and after any instance restart:
+
[source,ini]
----
# /etc/systemd/system/ptp-forward-gm.service
[Unit]
Description=Forward Grand Master parameters to the add-in ptp4l instances
After=ptp4l@nac.service ptp4l@aic1.service ptp4l@aic2.service

[Service]
Type=oneshot
ExecStart=/usr/local/bin/04-ptp-forward-GM.sh
----
+
[source,ini]
----
# /etc/systemd/system/ptp-forward-gm.timer
[Unit]
Description=Refresh forwarded Grand Master parameters

[Timer]
OnBootSec=2min
OnUnitActiveSec=5min

[Install]
WantedBy=timers.target
----
+
The effect is visible by querying an add-in `ptp4l` instance on the BC: before the forwarding

> it advertises its file defaults, and after it advertises the grandmaster's real values (which

Reviewer: ```suggestion
The effect is visible by querying an add-in `ptp4l` instance on the BC: Before forwarding, the node advertises its file defaults; after forwarding begins, it advertises the Grandmaster's actual values.(which
```

Author: Done — reworded with the before/after parallel structure. Used `Grandmaster` (one word) per the doc-wide terminology unification.
